qa/python_models/dlpack_test/model.py

```diff
@@ -99,28 +119,172 @@ def test_dlpack_gpu_tensors(self):
+            self.assertTrue(torch.equal(pytorch_tensor_dlpack, pytorch_tensor))
+            self.assertEqual(
+                pytorch_tensor.type(), pytorch_tensor_dlpack.type())
+            
 
     def test_dlpack_gpu_numpy(self):
         # DLPack tesnors that are in GPU cannot be converted to NumPy
         pytorch_tensor = torch.rand([100], dtype=torch.float16,
                                     device='cuda') * 100
         pb_tensor = pb_utils.Tensor.from_dlpack('tensor',
                                                 to_dlpack(pytorch_tensor))
+        # Make sure that `__dlpack_device__` works as expected
+        self.assertFalse(pb_tensor.is_cpu())
+        self.assertTrue(pytorch_tensor.is_cuda)
+        self.assertEqual(
+            pb_tensor.__dlpack_device__(), pytorch_tensor.__dlpack_device__())
+    
         with self.assertRaises(Exception) as e:
             pb_tensor.as_numpy()
         self.assertTrue(
             str(e.exception) ==
             'Tensor is stored in GPU and cannot be converted to NumPy.')
+    
+    def test_dlpack_cpu_numpy(self):
+        # Check compatibiity of PbTensor DLPack implementation
+        # with numpy
+        pytorch_tensor = torch.rand([100], dtype=torch.float16,
+                                    device='cpu') * 100
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', pytorch_tensor)
+        numpy_tensor_dlpack = np.from_dlpack(pb_tensor)
+        self.assertTrue(np.array_equal(
+            numpy_tensor_dlpack, pytorch_tensor.numpy()))
+        # Make sure that `__dlpack_device__` works as expected
+        self.assertTrue(pb_tensor.is_cpu())
+        self.assertFalse(pytorch_tensor.is_cuda)
+        self.assertEqual(
+            pb_tensor.__dlpack_device__(), pytorch_tensor.__dlpack_device__())
+    
+    def test_bool_datatype(self):
+        # [FIXME] pass bool_array directly to `pb_utils.Tensor.from_dlpack`,
+        # when numpy release supports DLPack bool type
+        bool_array = np.asarray([False, True])
+        bool_tensor = pb_utils.Tensor('tensor', bool_array)
+        bool_tensor_dlpack = pb_utils.Tensor.from_dlpack('tensor', bool_tensor)
+        self.assertTrue(
+            np.array_equal(bool_array, bool_tensor_dlpack.as_numpy()))
+    
+    def test_cuda_multi_stream(self):
+        # Test that external stream syncs with the default
+        # and pb_tensor has proper data
+        s1 = torch.cuda.Stream()
+        size = 5000
+        pytorch_tensor = torch.tensor([0,0,0,0], device='cuda')
+        expected_output = torch.tensor([2,2,2,2], device='cuda')
+        with torch.cuda.stream(s1):
+            matrix_a = torch.randn(size,size, device='cuda')
+            res = torch.matmul(matrix_a,matrix_a)
+            for _ in range(1000):
+                res = torch.matmul(res,matrix_a)
+            pytorch_tensor += torch.tensor([2,2,2,2], device='cuda')
+        
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', pytorch_tensor)
```

Comment:
does this work properly if we use `pb_utils.Tensor.from_dlpack('tensor', to_dlpack(pytorch_tensor))`?

Reply:
I'll check that